File: src/com/minhle/practiceleetcode/Triangle.java
package com.minhle.practiceleetcode;

import java.util.ArrayList;
import java.util.List;

public class Triangle {
    public int minimumTotal(List<List<Integer>> triangle) {
        List<Integer> minSum = new ArrayList<>();
        for (int i = 0; i <= triangle.size(); i++) {
            minSum.add(0);
        }
        for (int row = triangle.size() - 1; row >= 0; row --) {
            for (int column = 0; column < triangle.get(row).size(); column ++) {
                int n = triangle.get(row).get(column);
                minSum.set(column, n + Math.min(minSum.get(column), minSum.get(column + 1)));
            }
        }
        return minSum.get(0);
    }
}

File: src/com/minhle/practiceleetcode/Triangle.cpp
#include "AllNecessaryHeaders.h"

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        vector<vector<int>> dp(triangle.size(), vector<int>(triangle.back().size(), -1));
        return recursion(triangle, 0, 0, dp);
    }

    int recursion(const vector<vector<int>> &triangle, int i, int row, vector<vector<int>> &dp)
    {
        if (row == triangle.size())
        {
            return 0;
        }
        if (dp[i][row] != -1)
        {
            return dp[i][row];
        }
        dp[i][row] = triangle[row][i] + min(recursion(triangle, i, row + 1, dp), recursion(triangle, i + 1, row + 1, dp));
        return dp[i][row];
    }
};
